### ISSUE TITLE
Distortion when using the Perspective tool

### ISSUE THREAD

**Issue by fsenore** [Beep6581/RawTherapee#5558]
Windows 10 Pro.

The RawTherapee version is the official one for Windows:
Version: 5.7
Branch: releases
Commit: f335efe6a
Commit date: 2019-09-10
Compiler: x86_64-w64-mingw32-gcc\ 8.2.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.24.1
Lensfun: V0.3.2.0
Build type: Release
Build flags: -mwin32\ -m64\ -mthreads\ -msse2\ -std=c++11\ -mtune=generic\ -Werror=unused-label\ -Wall\ -Wuninitialized\ -Wcast-qual\ -Wno-deprecated-declarations\ -Wno-unused-result\ -fopenmp\ -Werror=unknown-pragmas\ -Wno-aggressive-loop-optimizations\ -DNDEBUG\ -O3\ -ftree-vectorize
Link flags: -m64\ -mthreads\ -static-libgcc\ -mtune=generic\ -s\ -O3
OpenMP support: ON
MMAP support: ON

I created this issue as suggested in this forum post where you can find complete information:

https://discuss.pixls.us/t/distortion-when-using-the-perspective-tool/15143

In short it looks like the Perspective correction tool in the Transform tab changes straight lines to bended ones while making them vertical. Maybe the cause is the heavy distortion correction applied by lensfun to certain lenses with large native distortion, like many micro 4/3 ones.

**Comment by Lawrence37** [comment]
It seems like the scaling for Auto-fill is applied in the wrong order. It should come after distortion correction.

Here's the output with Auto-fill enabled:
![P1240630_rt](https://user-images.githubusercontent.com/45837045/71605183-dff97380-2b1b-11ea-9701-6ce7b8e825aa.jpg)

and here's the output with Auto-fill disabled and scaled in GIMP:
![P1240630_gimp](https://user-images.githubusercontent.com/45837045/71605197-fa335180-2b1b-11ea-9181-b24f5e0f28ad.jpg)

**Comment by Lawrence37** [comment]
```diff
diff --git a/rtengine/iptransform.cc b/rtengine/iptransform.cc
index af513536e..805ad256f 100644
--- a/rtengine/iptransform.cc
+++ b/rtengine/iptransform.cc
@@ -253,15 +253,11 @@ bool ImProcFunctions::transCoord (int W, int H, const std::vector<Coord2D> &src,
     for (size_t i = 0; i < src.size(); i++) {
         double x_d = src[i].x, y_d = src[i].y;
 
-        if (pLCPMap && params->lensProf.useDist) {
-            pLCPMap->correctDistortion(x_d, y_d, 0, 0, ascale);
-        } else {
-            x_d *= ascale;
-            y_d *= ascale;
-        }
+        x_d -= w2;     // centering x coord & scale
+        x_d *= ascale;
+        y_d -= h2;     // centering y coord & scale
+        y_d *= ascale;
 
-        x_d += ascale * (0 - w2);     // centering x coord & scale
-        y_d += ascale * (0 - h2);     // centering y coord & scale
 
         if (needsPerspective()) {
             // horizontal perspective transformation
@@ -273,6 +269,10 @@ bool ImProcFunctions::transCoord (int W, int H, const std::vector<Coord2D> &src,
             y_d *= maxRadius * vpcospt / (maxRadius - y_d * vptanpt);
         }
 
+        if (pLCPMap && params->lensProf.useDist) {
+            pLCPMap->correctDistortion(x_d, y_d, 0, 0, 1);
+        }
+
         // rotate
         double Dx = x_d * cost - y_d * sint;
         double Dy = x_d * sint + y_d * cost;
@@ -935,15 +935,10 @@ void ImProcFunctions::transformGeneral(bool highQuality, Imagefloat *original, I
             double x_d = x;
             double y_d = y;
 
-            if (enableLCPDist) {
-                pLCPMap->correctDistortion(x_d, y_d, cx, cy, ascale); // must be first transform
-            } else {
-                x_d *= ascale;
-                y_d *= ascale;
-            }
-
-            x_d += ascale * centerFactorx; // centering x coord & scale
-            y_d += ascale * centerFactory; // centering y coord & scale
+            x_d += centerFactorx; // centering x coord & scale
+            x_d *= ascale;
+            y_d += centerFactory; // centering y coord & scale
+            y_d *= ascale;
 
             if (enablePerspective) {
                 // horizontal perspective transformation
@@ -955,6 +950,10 @@ void ImProcFunctions::transformGeneral(bool highQuality, Imagefloat *original, I
                 y_d *= maxRadius * vpcospt / (maxRadius - y_d * vptanpt);
             }
 
+            if (enableLCPDist) {
+                pLCPMap->correctDistortion(x_d, y_d, w2, h2, 1); // must be first transform
+            }
+
             // rotate
             const double Dxc = x_d * cost - y_d * sint;
             const double Dyc = x_d * sint + y_d * cost;
diff --git a/rtengine/rtlensfun.cc b/rtengine/rtlensfun.cc
index 7ffb9ad33..121f51e64 100644
--- a/rtengine/rtlensfun.cc
+++ b/rtengine/rtlensfun.cc
@@ -52,6 +52,9 @@ void LFModifier::correctDistortion(double &x, double &y, int cx, int cy, double
         return;
     }
 
+    x *= scale;
+    y *= scale;
+
     float pos[2];
     float xx = x + cx;
     float yy = y + cy;
@@ -67,8 +70,6 @@ void LFModifier::correctDistortion(double &x, double &y, int cx, int cy, double
         x -= cx;
         y -= cy;
     }
-    x *= scale;
-    y *= scale;
 }
```


With this patch, I get:
![P1240630_rt_patched](https://user-images.githubusercontent.com/45837045/71607428-44bdc980-2b2e-11ea-9049-0526315e9527.jpg)
@fsenore Is this the output you expect?

**Comment by fsenore** [comment]
@Lawrence37 Yes, the output is much better than before. Thank you for fixing this.
